README.md
# Gearman Job Server Chef Cookbook

[![Build Status](https://travis-ci.org/czeeb/gearman-job-server-cookbook.svg?branch=master)](https://travis-ci.org/czeeb/gearman-job-server-cookbook)
[![Dependency Status](https://gemnasium.com/czeeb/gearman-job-server-cookbook.svg)](https://gemnasium.com/czeeb/gearman-job-server-cookbook)
[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/czeeb/gearman-job-server-cookbook?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)

This is a [Chef] cookbook to manage [Gearman]

## Platforms
The following platforms and versions are tested and supported using [test-kitchen](http://kitchen.ci/)

* CentOS 6.6, 7.1
* Debian 7.8, 8.1
* Ubuntu 12.04, 14.04

Other versions that should work fine but are not testing in [test-kitchen].

* Amazon Linux 2015-03
* RHEL >= 6.6

## Tested Queue Types and Distros

An X indicates that the queue type for the distro is supported.  Different packages support different queue types because they had different compile time options and libraries used.

|              | Drizzle | Memcached | MySQL | Postgres | SQLite3 | Tokyo Cabinet |
|--------------|:-------:|:---------:|:-----:|:--------:|:-------:|:-------------:|
| CentOS 6.6   |         |           | X     | X        | X       | X             |
| CentOS 7.1   |         |           | X     | X        | X       | X             |
| Debian 7.8   |         | X         |       |          | X       |               |
| Debian 8.1   |         | X         | X     | X        | X       |               |
| Ubuntu 12.04 |         |           |       | X        | X       | X             |
| Ubuntu 14.04 |         | X         | X     | X        | X       |               |

## Attributes

Please see [gearman documentation](http://gearman.org/manual/job_server/#options) for what each of these options do.

* `default['gearman']['parameters']['backlog']` - Number of backlog connections for listen.
* `default['gearman']['parameters']['file-descriptors']` - Number of file descriptors to allow for the process.  Default is max allowed for users.
* `default['gearman']['parameters']['job-retries']` - Number of attempts to run the job before the job server removes it.  Default is no limit.
* `default['gearman']['parameters']['log-file']` - Log file to write errors and information to.
* `default['gearman']['parameters']['listen']` - Address the server should listen on. Default is INADDR_ANY.
* `default['gearman']['parameters']['port']` - Port the server should listen on.  Default is 4730.
* `default['gearman']['parameters']['threads']` - Number of I/O threads to use.  Default = 0
* `default['gearman']['parameters']['user']` - Switch to given user after startup

* `default['gearman']['queue-type']` - Persistent queue type to use. Valid options are libdrizzle, libmemcached, libsqlite3, libpq, mysql

* `default['gearman']['libdrizzle']['host']` - Host of server
* `default['gearman']['libdrizzle']['port']` - Port of server
* `default['gearman']['libdrizzle']['uds']` - Unix domain socket for server
* `default['gearman']['libdrizzle']['user']` - User name for authentication
* `default['gearman']['libdrizzle']['password']` - Password for authentication
* `default['gearman']['libdrizzle']['db']` - Database to use
* `default['gearman']['libdrizzle']['table']` - Table to use
* `default['gearman']['libdrizzle']['mysql']` - Set to true to use MySQL protocol

* `default['gearman']['libmemcached']['servers']` - List of memcached servers to use.

* `default['gearman']['libsqlite3']['db']` - Database file to use
* `default['gearman']['libsqlite3']['table']` - Table to use

* `default['gearman']['libpq']['conninfo']` - PostgreSQL connection information string
* `default['gearman']['libpq']['table']` - Table to use

* `default['gearman']['mysql']['host']` - Host of server
* `default['gearman']['mysql']['port']` - Port of server
* `default['gearman']['mysql']['user']` - User for authentication
* `default['gearman']['mysql']['password']` - Password for authentication
* `default['gearman']['mysql']['db']` - Database to use
* `default['gearman']['mysql']['table']` - Table to use

## Usage
#### gearman::default

e.g.
Just include `gearman` in your node's `run_list`:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[gearman]"
  ]
}
```

## Contributing

1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github to the development branch

## Helpful Stuff

* [Shared Serverspec Tests](https://gist.github.com/pburkholder/a9cdde0b33be8d5e3c08) - Explains why the serverspec tests are set up the way the are.

## License and Authors
Authors: Chris Zeeb <chris.zeeb@gmail.com> and [Contributors]

```text
Copyright:: 2015 Chris Zeeb <chris.zeeb@gmail.com>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```

[Chef]: https://www.chef.io
[Gearman]: http://gearman.org/
[Contributors]: https://github.com/czeeb/gearman-job-server-cookbook/graphs/contributors

CHANGELOG.md
# gearman-job-server-cookbook CHANGELOG

This file is used to list changes made in each version of the gearman cookbook.

## 1.0.0 / 2015-12-14

Rewrite of gearman cookbook

### Added

* Support for Amazon, CentOS, and RedHat
* `['gearman']['server']` options moved to `['gearman']['parameters']` 

### Changed

* Install from O/S packages instead of pulling package from github repository
* Added configuration options for:
  * backlog
  * file-descriptors
  * job-retries
  * log-file
  * listen
  * threads
  * queue-type
    * libdrizzle - all options
    * libmemcached - all options
    * libsqlite3 - all options
    * libpq - all options
    * mysql - all options
  

### Removed

* Installing gearman-ruby gem

## 0.0.1 / 2011-02-17

* Initial release.  The original cookbook can be found [here](https://github.com/gofullstack/gearman-cookbook).  The 1.0.0 release is a complete rewrite.


spec/spec_helper.rb
require 'chefspec'
require 'chefspec/berkshelf'

RSpec.configure do |config|
  config.log_level = :error
  config.platform = 'ubuntu'
  config.version = '14.04'
end

at_exit { ChefSpec::Coverage.report! }

spec/recipes/config_spec.rb
require_relative '../spec_helper.rb'

describe 'gearman::config' do
  context 'Debian Family' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new(platform: 'ubuntu', version: '14.04').converge(described_recipe)
    end

    it 'should create /etc/default/gearman-job-server' do
      expect(chef_run).to create_template('/etc/default/gearman-job-server')
    end
  end

  context 'EPEL Family' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new(platform: 'centos', version: '6.5').converge(described_recipe)
    end

    it 'should create /etc/sysconfig/gearmand' do
      expect(chef_run).to create_template('/etc/sysconfig/gearmand')
    end

    it 'should not create template[gearmand-init]' do
      expect(chef_run).to_not create_template('gearmand-init')
    end
  end

  context 'Debian 7.8' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new(platform: 'debian', version: '7.8').converge(described_recipe)
    end

    it 'should not create template[gearmand-init]' do
      expect(chef_run).to_not create_template('gearmand-init')
    end
  end

  context 'Debian 8.1' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new(platform: 'debian', version: '8.1').converge(described_recipe)
    end

    it 'should create template[gearmand-init]' do
      expect(chef_run).to create_template('gearmand-init')
    end
  end

  context 'Ubuntu 12.04' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new(platform: 'ubuntu', version: '12.04').converge(described_recipe)
    end

    it 'should create template[gearmand-init]' do
      expect(chef_run).to create_template('gearmand-init')
    end
  end

  context 'Ubuntu 14.04' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new(platform: 'ubuntu', version: '14.04').converge(described_recipe)
    end

    it 'should create template[gearmand-init]' do
      expect(chef_run).to create_template('gearmand-init')
    end
  end
end


spec/recipes/default_spec.rb
require_relative '../spec_helper.rb'

describe 'gearman::default' do
  let(:chef_run) { ChefSpec::SoloRunner.new.converge(described_recipe) }

  %w(
    gearman::install
    gearman::config
    gearman::service
  ).each do |recipe|
    it "includes recipe #{recipe}" do
      expect(chef_run).to include_recipe(recipe)
    end
  end
end


recipes/default.rb
#
# Cookbook Name:: gearman
# Recipe:: default
#
# Copyright (c) 2015 Chris Zeeb <chris.zeeb@gmail.com>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# TODO: Consistency pass.  gearman, gearmand, gearman-job-server.  All used for resources or reference.
include_recipe 'gearman::install'
include_recipe 'gearman::config'
include_recipe 'gearman::service'


recipes/install.rb
#
# Cookbook Name:: gearman
# Recipe:: install
#
# Copyright (c) 2015 Chris Zeeb <chris.zeeb@gmail.com>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# TODO: Add an option to install a specific version from source.

include_recipe 'apt' if node['platform_family'] == 'debian'

case node['platform_family']
when 'debian'
  package 'gearman-job-server'
  package 'gearman-tools'
when 'rhel'
  package 'epel-release'
  package 'gearmand'
end


recipes/service.rb
#
# Cookbook Name:: gearman
# Recipe:: service
#
# Copyright (c) 2015 Chris Zeeb <chris.zeeb@gmail.com>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe 'gearman::config'

# TODO: Investigate using the systemd chef cookbook instead
execute 'systemd-daemon-reload' do
  only_if { node['platform'] == 'debian' && node['platform_version'].to_f > 8.0 }
  command 'systemctl daemon-reload'
  action :run
  subscribes :run, 'template[gearman-init]', :delayed
end

service 'gearman-job-server' do
  action [:enable, :start]
  supports :restart => true, :start => true, :stop => true
  case node['platform_family']
  when 'debian'
    service_name 'gearman-job-server'
  when 'rhel'
    service_name 'gearmand'
  end
  subscribes :stop, 'template[gearman-config]', :delayed
  subscribes :start, 'template[gearman-config]', :delayed
  if node['platform'] == 'debian' && node['platform_version'].to_f > 8.0
    subscribes :stop, 'template[gearmand-init]', :delayed
    subscribes :start, 'template[gearmand-init]', :delayed
  end
end


recipes/config.rb
#
# Cookbook Name:: gearman
# Recipe:: config
#
# Copyright (c) 2015 Chris Zeeb <chris.zeeb@gmail.com>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Map parameters and discard nil k/v pairs
params = node['gearman']['parameters'].reject { |_k, v| v.nil? }.map { |k, v| "--#{k}=#{v}" }.join(' ')

# If a persistent queue-type is set, add the queue types parameters in and discard nil k/v pairs.
# TODO: Create connection string for libpq for user rather than them having to build it
if node['gearman']['parameters']['queue-type']
  queuetype = node['gearman']['parameters']['queue-type']
  queuetype = 'libpq' if queuetype == 'Postgres'

  # We want to ignore the libpq conninfo parameter for Ubuntu 12.04. It needs to go in the init script instead
  if node['platform'] == 'ubuntu' && node['platform_version'] == '12.04' && queuetype == 'libpq'
    params = params + ' ' + node['gearman'][queuetype].reject { |k, v| v.nil? || k == 'conninfo' }.map { |k, v| "--#{queuetype}-#{k}=#{v}" }.join(' ')
  else
    params = params + ' ' + node['gearman'][queuetype].reject { |_k, v| v.nil? }.map { |k, v| "--#{queuetype}-#{k}=#{v}" }.join(' ')
  end
end

# TODO: Add toggle for if gearman should be restarted when config changes?
template 'gearman-config' do
  source 'gearman-config.erb'
  case node['platform_family']
  when 'debian'
    path '/etc/default/gearman-job-server'
    variables(
      :param_word => 'PARAMS',
      :params     => params
    )
  when 'rhel'
    path '/etc/sysconfig/gearmand'
    variables(
      :param_word => 'OPTIONS',
      :params     => params
    )
  end
end

# TODO: Clean this up?  It's starting to look like a mess
template 'gearmand-init' do
  only_if { node['platform'] == 'ubuntu' || (node['platform'] == 'debian' && node['platform_version'].to_f > 8.0) }
  case node['platform']
  when 'ubuntu'
    case node['platform_version']
    when '12.04'
      path '/etc/init.d/gearman-job-server'
    when '14.04'
      path '/etc/init/gearman-job-server.conf'
    end
  when 'debian'
    path '/lib/systemd/system/gearman-job-server.service'
  end
  source 'gearmand.init.erb'
  variables(
    :libpq_conninfo => node['gearman']['libpq']['conninfo']
  )
end
